File: tcg/core/api/data_v2.py
"""Data v2 router — endpoints over the ``tcg_instruments_v2`` star schema.

Prefix ``/api/data-v2``. Parallel to the v1 ``/api/data`` router; reuses the
same read-only ``tcg_read`` pool via a distinct service
(:class:`DefaultMarketDataServiceV2`). Route ordering matters: the
``/continuous/*`` and ``/series/*`` routes are declared BEFORE any
``/objects/{object_id}`` path so a literal segment is never captured as an id
(the v1 catch-all gotcha).
"""

from __future__ import annotations

from fastapi import APIRouter, Depends, Query

from tcg.core.api._dates import parse_iso_range
from tcg.core.api.common import get_market_data_v2
from tcg.data.protocols import MarketDataServiceV2
from tcg.types.errors import DataNotFoundError, ValidationError
from tcg.types.market import AdjustmentMethod, ContinuousRollConfig, RollStrategy

router = APIRouter(prefix="/api/data-v2", tags=["data-v2"])


# --- Object browsing ---


@router.get("/objects")
async def list_objects(
    svc: MarketDataServiceV2 = Depends(get_market_data_v2),
) -> list[dict]:
    """List every v2 object (all kinds). The frontend groups by ``kind``."""
    objects = await svc.list_objects()
    return [
        {
            "object_id": o["object_id"],
            "kind": o["kind"],
            "symbol": o["symbol"],
            "name": o["name"],
            "cycle": o["cycle"],
            "underlying_object_id": o["underlying_object_id"],
        }
        for o in objects
    ]


# --- Continuous futures (declared BEFORE /objects/{object_id}) ---


@router.get("/continuous/futures/{object_id}/cycles")
async def get_future_cycles(
    object_id: int,
    svc: MarketDataServiceV2 = Depends(get_market_data_v2),
) -> dict:
    """Return available listing cycles for a future object."""
    cycles = await svc.get_future_cycles(object_id)
    return {"cycles": cycles}


@router.get("/continuous/futures/{object_id}")
async def get_continuous_future(
    object_id: int,
    strategy: str = Query("front_month", description="Roll strategy"),
    adjustment: str = Query(
        "none", description="Adjustment method: none, ratio, difference"
    ),
    cycle: str | None = Query(None, description="Listing cycle (informational in v2)"),
    roll_offset: int = Query(
        0, ge=0, le=365, description="Days before expiration to roll (0-365)"
    ),
    rank: int = Query(
        1,
        ge=1,
        le=12,
        description="NTH_NEAREST only: hold the rank-th nearest contract (1=front)",
    ),
    start: str | None = Query(None, description="Start date YYYY-MM-DD"),
    end: str | None = Query(None, description="End date YYYY-MM-DD"),
    svc: MarketDataServiceV2 = Depends(get_market_data_v2),
) -> dict:
    """Build a continuous futures series (v1-continuous-shape) for a v2 future."""
    try:
        roll_strategy = RollStrategy(strategy)
    except ValueError:
        raise ValidationError(
            f"Invalid strategy '{strategy}'. Must be one of: "
            f"{', '.join(e.value for e in RollStrategy)}"
        )
    try:
        adj_method = AdjustmentMethod(adjustment)
    except ValueError:
        raise ValidationError(
            f"Invalid adjustment '{adjustment}'. Must be one of: "
            f"{', '.join(e.value for e in AdjustmentMethod)}"
        )
    try:
        start_date, end_date = parse_iso_range(start, end)
    except ValueError as exc:
        raise ValidationError(str(exc)) from exc

    roll_config = ContinuousRollConfig(
        strategy=roll_strategy,
        adjustment=adj_method,
        cycle=cycle,
        roll_offset_days=roll_offset,
        rank=rank,
    )
    series = await svc.get_continuous_future(
        object_id, roll_config, start=start_date, end=end_date
    )
    if series is None:
        raise DataNotFoundError(f"No continuous series found for object {object_id}")
    return {
        "object_id": object_id,
        "collection": series.collection,
        "strategy": roll_config.strategy.value,
        "adjustment": roll_config.adjustment.value,
        "cycle": roll_config.cycle,
        "rank": roll_config.rank,
        "roll_dates": list(series.roll_dates),
        "contracts": list(series.contracts),
        "prices": {
            "dates": series.prices.dates.tolist(),
            "open": series.prices.open.tolist(),
            "high": series.prices.high.tolist(),
            "low": series.prices.low.tolist(),
            "close": series.prices.close.tolist(),
            "volume": series.prices.volume.tolist(),
        },
        # Flat mirrors (v1 parity) so a v1-style Chart consumer works unchanged.
        "dates": series.prices.dates.tolist(),
        "open": series.prices.open.tolist(),
        "high": series.prices.high.tolist(),
        "low": series.prices.low.tolist(),
        "close": series.prices.close.tolist(),
        "volume": series.prices.volume.tolist(),
    }


# --- Continuous options (v2-native settlement selection) ---


@router.get("/continuous/options/{object_id}")
async def get_continuous_options(
    object_id: int,
    criterion: str = Query(
        "strike", description="Selection criterion: strike | moneyness | delta"
    ),
    target: float | None = Query(
        None, description="Strike (absolute) or moneyness ratio"
    ),
    option_type: str = Query("put", description="call | put"),
    roll: str = Query("at_expiry", description="Roll rule (only at_expiry in v2)"),
    start: str | None = Query(None, description="Start date YYYY-MM-DD"),
    end: str | None = Query(None, description="End date YYYY-MM-DD"),
    svc: MarketDataServiceV2 = Depends(get_market_data_v2),
) -> dict:
    """Build a v2 continuous options settlement stream.

    ``criterion='delta'`` returns HTTP 400 (greeks unavailable in v2) — rejected
    outright, whether or not ``target`` is supplied. ``roll`` only supports
    ``at_expiry`` this round.
    """
    if roll != "at_expiry":
        raise ValidationError(
            f"Invalid roll '{roll}'. Only 'at_expiry' is supported in v2."
        )
    # Delta is rejected outright in v2 (no greeks) — return the friendly 400
    # regardless of whether ``target`` was given (an omitted target would
    # otherwise surface a generic 422 before this point).
    if criterion == "delta":
        raise ValidationError(
            "Delta-based selection is unavailable in Database v2: the v2 "
            "warehouse has no greeks (fact_greeks is empty). Use criterion "
            "'strike' or 'moneyness'."
        )
    if target is None:
        raise ValidationError(
            f"Query param 'target' is required for criterion {criterion!r} "
            "(absolute strike or moneyness ratio)."
        )
    try:
        start_date, end_date = parse_iso_range(start, end)
    except ValueError as exc:
        raise ValidationError(str(exc)) from exc

    result = await svc.get_continuous_options(
        object_id,
        criterion=criterion,
        target=target,
        option_type=option_type,
        start=start_date,
        end=end_date,
    )
    if not result.dates:
        raise DataNotFoundError(
            f"No option settlement data for object {object_id} "
            f"({option_type}, {criterion}={target}) in the given window"
        )
    return {
        "object_id": result.object_id,
        "criterion": result.criterion,
        "option_type": result.option_type,
        "target": target,
        "roll": roll,
        "spot_source": "underlying_future_front_close"
        if result.criterion == "moneyness"
        else None,
        "points": {
            "ts": list(result.dates),
            "value": list(result.values),
            "contract": list(result.contract_codes),  # per-date, 1:1 with ts
        },
        "roll_dates": list(result.roll_dates),
        "contracts": list(result.contracts),
    }


# --- Series facts (declared BEFORE /objects/{object_id}) ---


@router.get("/series/{serie_id}")
async def get_series(
    serie_id: int,
    start: str | None = Query(None, description="Start date YYYY-MM-DD"),
    end: str | None = Query(None, description="End date YYYY-MM-DD"),
    svc: MarketDataServiceV2 = Depends(get_market_data_v2),
) -> dict:
    """Read one serie's facts (fact table dispatched by ``serie.type``)."""
    try:
        start_date, end_date = parse_iso_range(start, end)
    except ValueError as exc:
        raise ValidationError(str(exc)) from exc
    return await svc.get_series(serie_id, start=start_date, end=end_date)


@router.get("/objects/{object_id}/facets")
async def get_object_facets(
    object_id: int,
    svc: MarketDataServiceV2 = Depends(get_market_data_v2),
) -> dict:
    """Filterable dimensions of one object — feeds the series filter form.

    Kept with the object routes, above the ``/objects/{object_id}`` catch-all.
    Unlike the single-segment routes above, declaration order is NOT
    load-bearing here: the catch-all compiles to ``objects/(?P<object_id>[^/]+)$``
    and ``[^/]+`` cannot span the ``/facets`` segment, so it can never capture
    this path whatever the order (measured — see
    ``test_facets_route_dispatches_to_the_facets_handler``). It would start to
    matter if that convertor were ever widened to ``{object_id:path}``.
    """
    return await svc.get_object_facets(object_id)


#: Filter enum domains, mirrored from the reader's whitelist so a bad value is
#: a clean 400 at the boundary rather than a DataAccessError from the adapter.
#: Duplicated rather than imported so the error message has a stable order (the
#: reader's are frozensets); ``test_route_enum_whitelists_match_the_readers``
#: pins them against each other so the copies cannot drift.
_SERIE_TYPE_VALUES = ("bar", "value", "greeks", "bbba", "any")
_FREQ_VALUES = ("1m", "daily", "any")
_OPTION_TYPE_VALUES = ("call", "put", "both")


@router.get("/objects/{object_id}/series")
async def list_object_series(
    object_id: int,
    expiration_min: str | None = Query(
        None, description="Earliest expiration YYYY-MM-DD"
    ),
    expiration_max: str | None = Query(
        None, description="Latest expiration YYYY-MM-DD"
    ),
    strike_min: float | None = Query(None, description="Strike lower bound"),
    strike_max: float | None = Query(None, description="Strike upper bound"),
    option_type: str = Query("both", description="call | put | both"),
    serie_type: str = Query("any", description="bar | value | greeks | bbba | any"),
    freq: str = Query("any", description="1m | daily | any"),
    skip: int = Query(0, ge=0),
    limit: int = Query(50, ge=1, le=500),
    svc: MarketDataServiceV2 = Depends(get_market_data_v2),
) -> dict:
    """One filtered, paginated page of an object's series.

    Kept with the object routes, above the ``/objects/{object_id}`` catch-all.
    As with ``/facets``, declaration order is NOT load-bearing for a
    two-segment path (``[^/]+`` cannot span the ``/series`` segment); the
    placement is for consistency. ``limit`` is capped at 500 — the same cap v1
    uses in ``tcg/core/api/data.py``.

    A filter that matches nothing returns 200 with an empty ``items`` and
    ``total: 0`` — a narrow filter is a result, not an error.
    """
    if serie_type not in _SERIE_TYPE_VALUES:
        raise ValidationError(
            f"Invalid serie_type {serie_type!r}. Must be one of: "
            f"{', '.join(_SERIE_TYPE_VALUES)}"
        )
    if freq not in _FREQ_VALUES:
        raise ValidationError(
            f"Invalid freq {freq!r}. Must be one of: {', '.join(_FREQ_VALUES)}"
        )
    if option_type not in _OPTION_TYPE_VALUES:
        raise ValidationError(
            f"Invalid option_type {option_type!r}. Must be one of: "
            f"{', '.join(_OPTION_TYPE_VALUES)}"
        )
    try:
        exp_min, exp_max = parse_iso_range(expiration_min, expiration_max)
    except ValueError as exc:
        raise ValidationError(str(exc)) from exc
    if exp_min is not None and exp_max is not None and exp_min > exp_max:
        raise ValidationError(
            f"expiration_min ({exp_min.isoformat()}) is after "
            f"expiration_max ({exp_max.isoformat()})"
        )
    if strike_min is not None and strike_max is not None and strike_min > strike_max:
        raise ValidationError(
            f"strike_min ({strike_min}) is greater than strike_max ({strike_max})"
        )

    return await svc.list_object_series(
        object_id,
        expiration_min=exp_min,
        expiration_max=exp_max,
        strike_min=strike_min,
        strike_max=strike_max,
        option_type=option_type,
        serie_type=serie_type,
        freq=freq,
        skip=skip,
        limit=limit,
    )


# --- Object detail (catch-all id route: declared LAST) ---


@router.get("/objects/{object_id}")
async def get_object_detail(
    object_id: int,
    svc: MarketDataServiceV2 = Depends(get_market_data_v2),
) -> dict:
    """Return ``{object}`` for one object — metadata only.

    Contracts and series are deliberately NOT here: they come from
    ``/objects/{id}/facets`` (aggregated dimensions) and
    ``/objects/{id}/series`` (filtered + paginated). Shipping them inline was
    38 239 859 bytes in ~36 s on object 12.
    """
    return await svc.get_object_detail(object_id)

File: frontend/src/api/dataV2.js
import { fetchApi } from './client';
import { classifyFetchError, FetchError } from '../utils/fetchError';

// ---------------------------------------------------------------------------
// Database v2 API client — explores the dwh star schema ``tcg_instruments_v2``
// via the ``/api/data-v2`` router. Mirrors the error-classification wrapper
// used in ``api/data.js`` and ``api/options.js`` (each file keeps its own local
// copy so none depends on another). All reads go through the shared
// ``fetchApi`` client, so ``API_BASE`` (``/api``) is prepended automatically.
// ---------------------------------------------------------------------------

async function fetchClassified(path, options = {}) {
  try {
    return await fetchApi(path, options);
  } catch (err) {
    // Let AbortError propagate unwrapped — callers check signal.aborted.
    if (err && err.name === 'AbortError') throw err;
    if (err && err.name === 'ApiError') {
      if (err.errorType === 'network_error') {
        const classified = classifyFetchError(new TypeError(err.message));
        throw new FetchError({ ...classified, cause: err });
      }
      const status = (err.details && err.details.status)
        || (err.errorType === 'not_found' ? 404 : null)
        || (err.errorType === 'validation' ? 400 : null)
        || (err.errorType === 'server_error' ? 500 : null)
        || null;
      if (status) {
        const classified = classifyFetchError(null, { status }, err.message);
        throw new FetchError({ ...classified, cause: err });
      }
    }
    const classified = classifyFetchError(err);
    throw new FetchError({ ...classified, cause: err });
  }
}

/**
 * GET /api/data-v2/objects
 * → [{ object_id, kind, symbol, name, cycle, underlying_object_id }]
 * All kinds (rate / index / future / option); the FE groups by ``kind``.
 */
export async function listObjectsV2({ signal } = {}) {
  const res = await fetchClassified('/data-v2/objects', { signal });
  // The endpoint returns a bare array; tolerate a ``{ objects: [...] }`` wrap.
  return Array.isArray(res) ? res : (res.objects || []);
}

/**
 * GET /api/data-v2/objects/{object_id}
 * → { object: { object_id, kind, symbol, name, cycle, underlying_object_id } }
 * Metadata only. Contracts and series come from `getObjectFacetsV2`
 * (aggregated dimensions) and `getObjectSeriesV2` (filtered + paginated) —
 * shipping them here was a 38 MB / ~36 s response.
 *
 * No component calls this today: the browser list already carries every field
 * the detail header renders, so `ObjectDetail` reads `object` from there rather
 * than re-fetching it. Kept because this module mirrors the v2 HTTP surface
 * one-for-one and the endpoint is live and cheap; see `useObjectDetailV2`.
 */
export async function getObjectDetailV2(objectId, { signal } = {}) {
  const res = await fetchClassified(
    `/data-v2/objects/${encodeURIComponent(objectId)}`,
    { signal },
  );
  return res;
}

/**
 * GET /api/data-v2/objects/{object_id}/facets
 * → { object_id, kind, expirations:[{expiration, contracts}], strike_min,
 *     strike_max, option_types:[...], serie_types:[{type, freq, series}],
 *     totals:{contracts, series} }
 * Cheap aggregate over the dimension tables — this is what the filter panel
 * is built from, so it never touches a fact table.
 */
export async function getObjectFacetsV2(objectId, { signal } = {}) {
  return fetchClassified(
    `/data-v2/objects/${encodeURIComponent(objectId)}/facets`,
    { signal },
  );
}

/**
 * Every key ``getObjectSeriesV2`` understands. Anything else is a caller bug —
 * see ``assertKnownSeriesFilters``. ``signal`` belongs here because the hook
 * calls the function as ``{ ...filters, signal }``.
 */
const SERIES_FILTER_KEYS = Object.freeze([
  'expirationMin', 'expirationMax', 'strikeMin', 'strikeMax',
  'optionType', 'serieType', 'freq', 'skip', 'limit', 'signal',
]);

/**
 * Reject unrecognised filter keys BEFORE any request is issued.
 *
 * Without this, an unknown key is silently dropped by the destructuring below
 * and the request goes out with the filter simply not applied — the caller gets
 * a plausible-looking page of the wrong rows and nothing anywhere says so. The
 * three ways that actually happens:
 *
 *   1. snake_case leaking in from the wire vocabulary:
 *      ``{ option_type: 'call' }`` → dropped; the chain comes back unfiltered.
 *   2. a plain typo: ``{ strikeMim: 6000 }`` → dropped.
 *   3. filters and options mis-slotted on the 3-arg hook:
 *      ``useObjectSeriesV2(12, { limit: 50, enabled: false })`` → passes the
 *      ``filters != null`` gate and FETCHES, while the author believes the hook
 *      is disabled. Easy slip, because ``useSeriesV2(serieId, { start, end,
 *      ...options })`` right next door folds params and options into one bag.
 *
 * A throw rather than a ``console.warn`` (the ``api/base.js`` precedent for
 * loud-but-non-fatal) because there is no valid reading of an unknown key: the
 * alternative to failing is returning confidently wrong data. It throws
 * synchronously, before ``fetchClassified``, so no request is issued — and it
 * is a plain ``TypeError``, not a ``FetchError``, because nothing was fetched.
 */
function assertKnownSeriesFilters(filters) {
  const unknown = Object.keys(filters).filter((k) => !SERIES_FILTER_KEYS.includes(k));
  if (unknown.length > 0) {
    throw new TypeError(
      `getObjectSeriesV2: unknown filter key(s) ${unknown.map((k) => `'${k}'`).join(', ')}. `
      + `Filters are camelCase and limited to: ${SERIES_FILTER_KEYS.join(', ')}. `
      + 'snake_case names (option_type, serie_type, strike_min, …) are the WIRE '
      + 'format produced by this function, never its argument format. If you meant '
      + 'to pass query options, they are the THIRD argument of useObjectSeriesV2.',
    );
  }
}

/**
 * Encode a strike bound, or return null when it is genuinely unset.
 *
 * ``''``/null/undefined mean "no bound" (the filter panel clears an input to
 * ''), but a non-finite number is a caller bug that MUST NOT reach the wire:
 * the backend declares ``strike_min: float | None = Query(None)``, and FastAPI
 * accepts the literal string ``"NaN"`` as ``nan`` with HTTP 200 (verified —
 * unlike ``"abc"``/``""``, which it rejects). So ``Number('1e')`` → NaN →
 * ``?strike_min=NaN`` → 200 with ``total: 0``, and the panel reports "no
 * series" for an object that has hundreds, with no error anywhere.
 */
function strikeBound(name, value) {
  if (value === undefined || value === null || value === '') return null;
  const num = Number(value);
  if (!Number.isFinite(num)) {
    // NOT JSON.stringify: it renders NaN and Infinity as "null", which is the
    // single most misleading thing this message could say.
    const shown = typeof value === 'string' ? `'${value}'` : String(value);
    throw new TypeError(
      `getObjectSeriesV2: ${name} must be a finite number, received ${shown}. `
      + 'A NaN bound is accepted by the backend with HTTP 200 and silently '
      + 'matches nothing, so it is rejected here instead.',
    );
  }
  // Emit the parsed number, not the raw input: validating `Number(value)` and
  // then sending `String(value)` would let '  6000  ' through with its spaces.
  // Byte-identical to the raw form for every well-formed value (6000 → '6000',
  // 0 → '0', '6000' → '6000').
  return String(num);
}

/**
 * GET /api/data-v2/objects/{object_id}/series?<filters>&skip&limit
 * → { items:[{serie_id, contract_id, type, freq, source, contract_code,
 *     expiration, strike, option_type}], total, skip, limit }
 * Contract metadata arrives joined, so no contract_id → contract map is needed
 * client-side. An empty `items` with `total: 0` is a normal answer.
 *
 * Domain reminder (v2, NOT v1): ``optionType`` is call|put|both, ``serieType``
 * is bar|value|greeks|bbba|any, ``freq`` is 1m|daily|any. ``limit`` defaults to
 * 50 server-side and is capped at 500 — out of range comes back as HTTP 400.
 *
 * Omission is meaningful: a filter left unset is NOT sent, so the backend
 * applies its own default (option_type=both / serie_type=any / freq=any)
 * rather than receiving an empty string.
 *
 * Throws ``TypeError`` (before issuing anything) on an unknown filter key or a
 * non-finite strike bound — both are silently-wrong-results bugs otherwise.
 */
export async function getObjectSeriesV2(objectId, filters = {}) {
  assertKnownSeriesFilters(filters);
  const {
    expirationMin,
    expirationMax,
    strikeMin,
    strikeMax,
    optionType,
    serieType,
    freq,
    skip,
    limit,
    signal,
  } = filters;
  const params = new URLSearchParams();
  if (expirationMin) params.set('expiration_min', expirationMin);
  if (expirationMax) params.set('expiration_max', expirationMax);
  const strikeMinParam = strikeBound('strikeMin', strikeMin);
  if (strikeMinParam !== null) params.set('strike_min', strikeMinParam);
  const strikeMaxParam = strikeBound('strikeMax', strikeMax);
  if (strikeMaxParam !== null) params.set('strike_max', strikeMaxParam);
  if (optionType) params.set('option_type', optionType);
  if (serieType) params.set('serie_type', serieType);
  if (freq) params.set('freq', freq);
  if (skip) params.set('skip', String(skip));
  if (limit) params.set('limit', String(limit));
  const query = params.toString() ? `?${params}` : '';
  return fetchClassified(
    `/data-v2/objects/${encodeURIComponent(objectId)}/series${query}`,
    { signal },
  );
}

/**
 * GET /api/data-v2/series/{serie_id}?start&end
 * → { serie_id, type, fields:[...], points:{ ts:[...], <field>:[...] } }
 * ``type`` dispatches the chartable field set (bar→OHLCV+OI, value→value,
 * greeks→…, bbba→…).
 */
export async function getSeriesV2(serieId, { start, end, signal } = {}) {
  const params = new URLSearchParams();
  if (start) params.set('start', start);
  if (end) params.set('end', end);
  const query = params.toString() ? `?${params}` : '';
  const res = await fetchClassified(
    `/data-v2/series/${encodeURIComponent(serieId)}${query}`,
    { signal },
  );
  return res;
}

/**
 * GET /api/data-v2/continuous/futures/{object_id}
 *     ?strategy&adjustment&cycle&roll_offset&rank&start&end
 * → v1-continuous shape family: { dates, open, high, low, close, volume,
 *   roll_dates, contracts } (reviewer: confirm BE returns v1-shaped price
 *   arrays keyed ``dates``/``close`` etc., not a nested ``prices`` object).
 * Mirrors ``getContinuousSeries`` in api/data.js (same param encoding).
 */
export async function getContinuousFuturesV2(objectId, {
  strategy = 'front_month',
  adjustment = 'none',
  cycle,
  rollOffset,
  rank,
  start,
  end,
} = {}) {
  const params = new URLSearchParams();
  params.set('strategy', strategy);
  params.set('adjustment', adjustment);
  if (cycle) params.set('cycle', cycle);
  if (rollOffset > 0) params.set('roll_offset', String(rollOffset));
  if (rank > 1) params.set('rank', String(rank));
  if (start) params.set('start', start);
  if (end) params.set('end', end);
  const res = await fetchClassified(
    `/data-v2/continuous/futures/${encodeURIComponent(objectId)}?${params}`,
  );
  return res;
}

/** GET /api/data-v2/continuous/futures/{object_id}/cycles → available cycles. */
export async function getV2FuturesCycles(objectId) {
  const res = await fetchClassified(
    `/data-v2/continuous/futures/${encodeURIComponent(objectId)}/cycles`,
  );
  return Array.isArray(res) ? res : (res.cycles || []);
}

/**
 * GET /api/data-v2/continuous/options/{object_id}
 *     ?criterion=strike|moneyness&target&option_type=call|put&roll=at_expiry
 *     &start&end
 * → { points:{ ts, value }, roll_dates, contracts, spot_source? }
 * ``criterion=delta`` is rejected by the BE (422/400 "greeks unavailable in
 * v2"); the FE also greys the Delta option so this path is unreachable from UI.
 */
export async function getContinuousOptionsV2(objectId, {
  criterion = 'strike',
  target,
  optionType = 'put',
  roll = 'at_expiry',
  start,
  end,
} = {}) {
  const params = new URLSearchParams();
  params.set('criterion', criterion);
  if (target !== undefined && target !== null && target !== '') {
    params.set('target', String(target));
  }
  params.set('option_type', optionType);
  params.set('roll', roll);
  if (start) params.set('start', start);
  if (end) params.set('end', end);
  const res = await fetchClassified(
    `/data-v2/continuous/options/${encodeURIComponent(objectId)}?${params}`,
  );
  return res;
}
